### PR TITLE
Manual example of generated trait code for Kotlin

### PR DESCRIPTION
**Pull request by emarteca**
This is a manual example of what the generated Kotlin should look like for trait usage with Diplomat, for an example trait `TesterTrait`. It also includes an example usage of this generated code.

[This commit](https://github.com/rust-diplomat/diplomat/commit/482f978aa79a3d40e53c2e1853af87d7227bb83e) has the example code. 

Feedback welcome!

### CODE REVIEW

**Comment by emarteca** [comment]
+ @jcrist1

File: example/simple_testing/kotlin/src/main/kotlin/dev/diplomattest/simple_testing/Main.kt [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    fun Wrapper_test_with_trait(diplomatTraitWrapper: DiplomatTrait_TesterTrait_Wrapper_Native, i: Int): Int
+}
+
+// the Kotlin interface that you'll extend to have access to the Rust trait `TesterTrait`
+interface DiplomatTraitInterface_TesterTrait {
+    fun test_trait_fn(x: Int): Int;
+    fun test_void_trait_fn(): Unit;
+}
```
Comment (jcrist1) [inline example/simple_testing/kotlin/src/main/kotlin/dev/diplomattest/simple_testing/Main.kt:100]: If I were a consumer of this API I'd prefer to use less boilerplate in names and style conforming to the host language. For the internal stuff it doesn't matte but is there any reason we couldn't do this?

```suggestion
interface TesterTrait {
    fun testTraitFn(x: Int): Int;
    fun testVoidTrait_fn(): Unit;
}
```
It feels like it be much more ergonomic to the person whose using the library on the kotlin side.

File: example/simple_testing/kotlin/src/main/kotlin/dev/diplomattest/simple_testing/Main.kt [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    internal var vtable: DiplomatTrait_TesterTrait_VTable_Native
+        = DiplomatTrait_TesterTrait_VTable_Native();
+  
+    // Define the fields of the struct
+    override fun getFieldOrder(): List<String> {
+        return listOf("data_", "vtable")
+    }
+}
+
+class DiplomatCallback_TesterTrait_Wrapper internal constructor (
```
Comment (jcrist1) [inline example/simple_testing/kotlin/src/main/kotlin/dev/diplomattest/simple_testing/Main.kt:48]: I'd expect something like this
```suggestion
class Wrapper internal constructor (
```

File: example/simple_testing/kotlin/src/main/kotlin/dev/diplomattest/simple_testing/Main.kt [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            }
+            val vtable = DiplomatTrait_TesterTrait_VTable_Native()
+            vtable.run_test_trait_fn_callback = test_trait_fn;
+            vtable.run_test_void_trait_fn_callback = test_void_trait_fn;
+            val native_wrapper = DiplomatTrait_TesterTrait_Wrapper_Native();
+            native_wrapper.vtable = vtable;
+            return DiplomatCallback_TesterTrait_Wrapper(native_wrapper);
+        }
+
+        fun test_trait_fn(trait_wrap: DiplomatCallback_TesterTrait_Wrapper, i: Int): Int {
```
Comment (jcrist1) [inline example/simple_testing/kotlin/src/main/kotlin/dev/diplomattest/simple_testing/Main.kt:77]:
```suggestion
        fun testTraitFn(traitWrap: Wrapper, i: Int): Int {
```

**Review by jcrist1 — COMMENTED** [review]
I just have a couple of proposed style changes so far. I'll try and take a deeper look on the weekend.

**Comment by jcrist1** [comment]
@emarteca I wasn't expecting that there'd be a possibility of generic programming with Diplomat, so this will be  a cool feature. The ownership issue is still a bit mindbending, but I don't think it's necessary for this example.

File: example/simple_testing/kotlin/src/main/kotlin/dev/diplomattest/simple_testing/Main.kt [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+interface DiplomatTraitInterface_TesterTrait {
+    fun test_trait_fn(x: Int): Int;
+    fun test_void_trait_fn(): Unit;
+}
+
+// ------------------------------------------------------------------------------------------------ user written code below
+
+// class implementing the interface -- this is where you add custom functionality
+// for the trait methods
+class MyTesterTraitImpl: DiplomatTraitInterface_TesterTrait {
```
Comment (emarteca) [inline example/simple_testing/kotlin/src/main/kotlin/dev/diplomattest/simple_testing/Main.kt:106]: add a `to_native` or something similar thats equivalent to fromTraitObj (for ergs)

**Review by jcrist1 — COMMENTED** [review]
I have one naming proposal, and then a bit more of a question @emarteca:

In general I still find the idea of ownership crossing over a little bit daunting, i.e. that we pass in something that implements this trait and then Rust would clean it up. For a managed language like Kotlin, it feels like we're prone to deleting something that might still be used somewhere else. I can't immediately tell but it seems to me that an` impl Trait` is just a type like any other, and it can have borrows. In kotlin we would almost always want to take `& impl Trait` or maybe `&mut impl Trait` and as a rust user I'd feel more comfortable knowing I'm not taking ownership when I don't need to. Is that possible?

**Comment by Manishearth** [comment]
> In general I still find the idea of ownership crossing over a little bit daunting, i.e. that we pass in something that implements this trait and then Rust would clean it up. For a managed language like Kotlin, it feels like we're prone to deleting something that might still be used somewhere else. I can't immediately tell but it seems to me that an` impl Trait` is just a type like any other, and it can have borrows. In kotlin we would almost always want to take `& impl Trait` or maybe `&mut impl Trait` and as a rust user I'd feel more comfortable knowing I'm not taking ownership when I don't need to. Is that possible?

So I don't think ownership would actually "cross over". The Rust `impl Trait` would contain a destructor that would do globalref shenanigans to "deregister" the object on the other side, but it wouldn't actually delete it.

This is the same as what we plan to do for callbacks, though it's not implemented yet.

basically, `impl Trait` doesn't imply anything about ownership, since the thing behind the `impl Trait` can be conceptually reference counted/garbage collected.

C++ can have a proper cleanup dtor if it wishes.

File: example/simple_testing/kotlin/src/main/kotlin/dev/diplomattest/simple_testing/Main.kt (hunk 1 above)
Comment (emarteca) [reply example/simple_testing/kotlin/src/main/kotlin/dev/diplomattest/simple_testing/Main.kt:100]: Great point! we'll do that.

File: core/src/ast/traits.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+use super::docs::Docs;
+use super::{Attrs, Ident, LifetimeEnv, Param, PathTrait, TraitSelfParam, TypeName};
+
+/// A struct declaration in an FFI module that is not opaque.
+#[derive(Clone, PartialEq, Eq, Hash, Serialize, Debug)]
+#[non_exhaustive]
+pub struct Trait {
+    pub name: Ident,
+    pub lifetimes: LifetimeEnv,
+    pub fcts: Vec<TraitMethod>,
```
[the remark this answers is not in the document]
Comment (emarteca) [reply core/src/ast/traits.rs:12]: Sounds good, @Manishearth had the same comment :-)

**Comment by emarteca** [comment]
Thanks for your comments! Indeed, our planned ownership model here is that ownership of the trait objects is not transferred -- as Manish says, the Rust `impl Trait` object will hold a JNA GlobalRef to the Kotlin object passed in. This will make sure the Rust object isn't dropped while the corresponding Kotlin object still exists. Then, in the Rust destructor we'll drop the GlobalRef, so that when it is garbage collected on the Kotlin side it can actually get removed.

I'm going to do this first thing next week, we're tracking it in [this issue](https://github.com/rust-diplomat/diplomat/issues/648) because we also need it for proper callback support.

**Comment by emarteca** [comment]
Closing this PR since it is now obsolete; we merged https://github.com/rust-diplomat/diplomat/pull/689 with the code generation for traits in Kotlin

**Pull request closed without merging**